### PR TITLE
[IOTDB-5466] Refactor RaftLog disk monitor to avoid unnecessary snapshots

### PR DESCRIPTION
**Pull request by SzyWilliam**
see https://issues.apache.org/jira/browse/IOTDB-5466

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java [hunk 1]
```diff
@@ -265,4 +271,52 @@ public static void initRatisConfig(RaftProperties properties, RatisConfig config
...
+      }
+    }
+
+    private final ConcurrentHashMap<File, DirectoryState> directoryMap = new ConcurrentHashMap<>();
+
+    long updateAndGetDirectorySize(File dir) {
+      final DirectoryState prev = directoryMap.computeIfAbsent(dir, d -> new DirectoryState());
+      List<Path> latest;
+      try (Stream<Path> files = Files.list(dir.toPath())) {
+        latest = files.filter(p -> p.getFileName().startsWith("log_inprogress"))  // filter out the in progress ratis log
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:297]: Remove Magic String

Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java]: fixed. It is now a static predicate.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java [hunk 2]
```diff
@@ -723,11 +722,19 @@ private void triggerSnapshotByCustomize() {
...
 
       if (currentDirLength >= triggerSnapshotFileSize) {
+        final StringBuilder allFileNamesBuilder = new StringBuilder();
+        monitor.getFilesUnder(currentDir).forEach(p -> allFileNamesBuilder.append(p).append(","));
+        logger.info(
+            "{}: take snapshot for region {}, current dir size {}, files {}",
+            this,
+            raftGroupId,
+            currentDirLength,
+            allFileNamesBuilder);
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:736]: whether this field will be too long?

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java [hunk 3]
```diff
@@ -265,4 +276,52 @@ public static void initRatisConfig(RaftProperties properties, RatisConfig config
     RaftServerConfigKeys.Rpc.setFirstElectionTimeoutMax(
         properties, config.getRpc().getFirstElectionTimeoutMax());
   }
+
+  /**
+   * Monitoring Ratis RaftLog total size. It will memorize the state of all files in the last update
+   * and calculates the diff incrementally each run.
+   */
+  static class RatisLogSizeMonitor {
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:284]: It feels like this class could be a separate tool class

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java [hunk 4]
```diff
@@ -265,4 +276,52 @@ public static void initRatisConfig(RaftProperties properties, RatisConfig config
...
+
+    private final ConcurrentHashMap<File, DirectoryState> directoryMap = new ConcurrentHashMap<>();
+
+    long updateAndGetDirectorySize(File dir) {
+      final DirectoryState prev = directoryMap.computeIfAbsent(dir, d -> new DirectoryState());
+      List<Path> latest;
+      try (Stream<Path> files = Files.list(dir.toPath())) {
+        latest = files.filter(isOpenSegment).collect(Collectors.toList());
+      } catch (IOException e) {
+        RatisConsensus.logger.warn(
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:304]: use it's solitary logger？

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java [hunk 5]
```diff
@@ -265,4 +276,52 @@ public static void initRatisConfig(RaftProperties properties, RatisConfig config
...
+
+  /**
+   * Monitoring Ratis RaftLog total size. It will memorize the state of all files in the last update
+   * and calculates the diff incrementally each run.
+   */
+  static class RatisLogSizeMonitor {
+
+    private static final class DirectoryState {
+      private long size = 0;
+      private List<Path> memorizedFiles = Collections.emptyList();
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:288]: Use `set` to make `contains` faster?

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java [hunk 6]
```diff
@@ -265,4 +276,52 @@ public static void initRatisConfig(RaftProperties properties, RatisConfig config
...
+
+      private void update(long size, List<Path> latest) {
+        this.size = size;
+        this.memorizedFiles = latest;
+      }
+    }
+
+    private final ConcurrentHashMap<File, DirectoryState> directoryMap = new ConcurrentHashMap<>();
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:296]: Why need `ConcurrentHashMap` as there is only one thread?

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java (hunk 2 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/RatisConsensus.java:736]: I'll just print how many files contained.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java (hunk 5 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:288]: Fixed

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java (hunk 6 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:296]: Changed to HashMap

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java (hunk 4 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:304]: Fixed

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java (hunk 3 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/Utils.java:284]: Moved to Utils package

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM

**Review by BUAAserein — APPROVED** [review]
LGTM~

**Pull request merged**